### PR TITLE
Bug 1823482 - Handle opted-out Ad IDs

### PR DESCRIPTION
**Pull request by fbertsch**
Checklist for reviewer:

- [ ] Commits should reference a bug or github issue, if relevant (if a bug is referenced, the pull request should include the bug number in the title).
- [ ] If the PR comes from a fork, trigger integration CI tests by running the [Push to upstream workflow](https://github.com/mozilla/bigquery-etl/actions/workflows/push-to-upstream.yml) and provide the `<username>:<branch>` of the fork as parameter. The parameter will also show up
in the logs of the `manual-trigger-required-for-fork` CI task together with more detailed instructions.
- [ ] If adding a new field to a query, ensure that the schema and dependent downstream schemas have been updated.
- [ ] When adding a new derived dataset, ensure that data is not available already (fully or partially) and recommend extending an existing dataset in favor of creating new ones. Data can be available in the [bigquery-etl repository](https://github.com/mozilla/bigquery-etl), [looker-hub](https://github.com/mozilla/looker-hub) or in [looker-spoke-default](https://github.com/mozilla/looker-spoke-default/tree/e1315853507fc1ac6e78d252d53dc8df5f5f322b).

For modifications to schemas in restricted namespaces (see [`CODEOWNERS`](https://github.com/mozilla/bigquery-etl/blob/main/CODEOWNERS)):
- [ ] Follow the [change control procedure](https://docs.google.com/document/d/1TTJi4ht7NuzX6BPG_KTr6omaZg70cEpxe9xlpfnHj9k/edit#heading=h.ttegrcfy18ck)

### CODE REVIEW

File: sql/moz-fx-data-shared-prod/udf/pseudonymize_ad_id/udf.sql [hunk 1]
```diff
@@ -0,0 +1,23 @@
+-- pseudonymize_ad_ids: Only for use in Ad ID pipeline
+CREATE OR REPLACE FUNCTION udf.pseudonymize_ad_id(hashed_ad_id STRING, key BYTES)
+RETURNS STRING AS (
+  IF( -- These two values represent the hashed all-0 Ad IDs, see bug 1823482
+    hashed_ad_id IN (
+      "q7RZx9leE4AHjn9DRQR8kLtjO0FNwmj/61ECGxFinno=",
+      "ksEEnM6/esBxl2myqrVLo31ePBoQZZnkCm81vefaF90="
+    ),
+    NULL,
+    TO_HEX(`moz-fx-data-shared-prod`.udf.hmac_sha256(key, CAST(ad_id AS BYTES)))
```
Comment (scholtzan) [inline sql/moz-fx-data-shared-prod/udf/pseudonymize_ad_id/udf.sql:10]:
```suggestion
    TO_HEX(`moz-fx-data-shared-prod`.udf.hmac_sha256(key, CAST(hashed_ad_id AS BYTES)))
```

File: sql/moz-fx-data-shared-prod/udf/pseudonymize_ad_id/README.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+
+<!--
+Warning: Only for use in the Ad ID pipeline.
+
+This UDF uses a key to pseduonymize Ad IDs,
+and handles the all-0 Ad IDs we get from
+opted-out clients.
+-->
```
Comment (scholtzan) [inline sql/moz-fx-data-shared-prod/udf/pseudonymize_ad_id/README.md:8]:
```suggestion
Warning: Only for use in the Ad ID pipeline.

This UDF uses a key to pseduonymize Ad IDs,
and handles the all-0 Ad IDs we get from
opted-out clients.
```

I don't think it is strictly necessary to have this as a comment. Though these docs aren't being rendered anywhere, so I guess it doesn't really matter

**Review by scholtzan — APPROVED** [review]
r+wc

File: sql/moz-fx-data-shared-prod/udf/pseudonymize_ad_id/udf.sql (hunk 1 above)
Comment (fbertsch) [reply sql/moz-fx-data-shared-prod/udf/pseudonymize_ad_id/udf.sql:10]: hmm these tests must have not run or they'd have failed

Comment (scholtzan) [reply sql/moz-fx-data-shared-prod/udf/pseudonymize_ad_id/udf.sql:10]: CI did fail

Comment (fbertsch) [reply sql/moz-fx-data-shared-prod/udf/pseudonymize_ad_id/udf.sql:10]: So it did!

**Pull request merged**